cpp/include/cudf/strings/detail/convert/string_to_float.cuh

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/strings/detail/convert/is_float.cuh>
+#include <cudf/strings/string_view.cuh>
+
+#include <cmath>
+#include <limits>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+/**
+ * @brief This function converts the given string into a
+ * floating point double value.
+ *
+ * This will also map strings containing "NaN", "Inf", etc.
+ * to the appropriate float values.
+ *
+ * This function will also handle scientific notation format.
+ */
+__device__ inline double stod(string_view const& d_str)
+{
+  const char* in_ptr = d_str.data();
+  const char* end    = in_ptr + d_str.size_bytes();
+  if (end == in_ptr) return 0.0;
+  double sign{1.0};
+  if (*in_ptr == '-' || *in_ptr == '+') {
+    sign = (*in_ptr == '-' ? -1 : 1);
+    ++in_ptr;
+  }
+
+#ifndef CUDF_JIT_UDF
+  constexpr double infinity      = std::numeric_limits<double>::infinity();
+  constexpr uint64_t max_holding = (std::numeric_limits<uint64_t>::max() - 9L) / 10L;
+#else
+  constexpr double infinity      = (1.0 / 0.0);
```

Comment:
Wow, will this have any side effect, or the result depends on the compiler flag?

Reply:
I meant, can we guarantee that the runtime will not throw exception here?

Reply:
It is only for the JIT compiler and likely temporary once #11319 is sorted out.